pageserver/src/http/routes.rs

```diff
@@ -354,6 +355,32 @@ async fn status_handler(
     json_response(StatusCode::OK, StatusResponse { id: config.id })
 }
 
+async fn reload_auth_validation_keys_handler(
```

Comment:
> Can we just restart pageserver?

We can, but it will mean downtime, at least before HA work. With HA work, it will not be directly responsible for downtime, but it will reduce the number of pageservers that can respond so it will degrade our service quality for the time of the key rollout. Such a degradation would already be a regular occurence due to deploys so one could say that it's an okay cost. Then #4897 would not be needed.

>  then it looks like a more generic 'live-reload pageserver configuration'

A generic live reload feature was one of the alternatives I laid out [here](https://github.com/neondatabase/neon/issues/4897#issuecomment-1668597469), but I agree with John that something specific just for reloading the specified jwt path (directory or file) is better because it's easier to implement.

Note that this PR does not allow a change of the actual parts stored in toml, namely the path. That path remains the same, the only thing that is reloaded is the config that the path points to (as we don't put the jwt keys into the toml file directly).

The generic ability for the pageserver to reload arbitrary configuration is a big feature. We probably would only support a subset of config params, and would have to start a discussion about what that subset should be, probably influenced by the use cases. To me it feels that there is few such use cases where, but I might be wrong.

Right now we *need* only the jwt tokens. And this live reload ability is just a clone away from being broken (at least for easily cloneable types like ints, paths or strings. People might do such clones without noticing, so I wonder how we should preserve the live reload ability after we add it. 

The other issue is how to manage the ability to change the config from a central point and at the same time, be able to access it from multiple threads.

>  having a specific http API just to reload keys, looks a bit strange to me, but maybe I miss something.

it is easier to implement on the pageserver side, as we already have multiple http apis, compared to SIGHUP. But I'm fine with SIGHUP as well. I can look into it.

Reply:
> it is easier to implement on the pageserver side, as we already have multiple http apis, compared to SIGHUP. But I'm fine with SIGHUP as well. I can look into it.

No, I'm fine with http API too if generic live-reload is more complicated and you wanted to narrow down the scope. Thanks for the explanation

I still have two comments

1. It'd be nice to add an OpenAPI spec for this new API, so it was clear how to use

2. Shouldn't it be just empty-body `POST /reload_auth_validation_keys`? If path is set in the toml config, then ps has to only re-read the keys directory, right? From the python test it looks like that, but in Rust code it's read from the request?

Reply:
> It'd be nice to add an OpenAPI spec for this new API

Good point, I will.

> If path is set in the toml config, then ps has to only re-read the keys directory, right?

Yes, precisely. The API call only does the reload, but nothing else. You cannot specify any directory or such, and I've been advised against adding the ability to specify a directory by @ tychoish.

Reply:
So, we'll have to update the deploy scripts to run this API?

Reply:
> we'll have to update the deploy scripts to run this API?

if you want to call the API, sure. but it's not obligatory, only if we want to support key rollover without needing restarts (which is a good idea IMO).